core-plugins/src/main/java/co/cask/hydrator/plugin/batch/source/XMLReaderBatchSource.java

```diff
@@ -0,0 +1,302 @@
+          } else {
+            processedFiles.add(Bytes.toString(keyValue.getKey()));
+          }
+        }
+      }
+      //File name use by BatchXMLFileFilter to filter already processed files.
+      conf.set(XMLInputFormat.XML_INPUTFORMAT_PROCESSED_FILES,
+               GSON.toJson(processedFiles, ARRAYLIST_PREPROCESSED_FILES));
+    }
+  }
+
+  @Override
+  public void transform(KeyValue<LongWritable, Object> input, Emitter<StructuredRecord> emitter) throws Exception {
+    Map<String, String> xmlRecord = (Map<String, String>) input.getValue();
+    Set<String> keySet = xmlRecord.keySet();
+    Iterator<String>  itr = keySet.iterator();
+    String fileName = Iterators.getOnlyElement(itr);
+    String record = xmlRecord.get(fileName);
+
+    StructuredRecord output = StructuredRecord.builder(DEFAULT_XML_SCHEMA)
+      .set("offset", input.getKey().get())
+      .set("filename", fileName)
+      .set("record", record)
+      .build();
+    emitter.emit(output);
+  }
+
+  @Override
+  public void onRunFinish(boolean succeeded, BatchSourceContext context) {
+    super.onRunFinish(succeeded, context);
+    try {
+      FileStatus[] status = fileSystem.listStatus(tempDirectoryPath);
+      long processingTime = new Date().getTime();
+      Path[] paths = FileUtil.stat2Paths(status);
+      if (paths != null && paths.length > 0) {
+        for (Path path : paths) {
+          try (FSDataInputStream input = fileSystem.open(path)) {
+            String key = input.readUTF();
+            processedFileTrackingTable.write(Bytes.toBytes(key), Bytes.toBytes(processingTime));
+          }
+        }
+      }
+    } catch (IOException exception) {
+      LOG.error("IOException occurred while reading temp directory path : " + exception.getMessage());
+    }
+  }
+
+  /**
+   * Config class that contains all the properties needed for the XML Reader.
+   */
+  public static class XMLReaderConfig extends ReferencePluginConfig {
+    @Description("Path to file(s) to be read. If a directory is specified, terminate the path name with a \'/\'.")
+    private final String path;
+
+    @Nullable
+    @Description("Pattern to select specific file(s)." +
+      "Example - " +
+      "1. Use '^' to select file with name start with 'catalog', like '^catalog'." +
+      "2. Use '$' to select file with name end with 'catalog.xml', like 'catalog.xml$'." +
+      "3. Use '*' to select file with name contains 'catalogBook', like 'catalogBook*'.")
+    private final String pattern;
+
+    @Description("Node path to emit individual event from the schema. " +
+      "Example - '/book/price' to read only price under the book node")
+    private final String nodePath;
+
+    @Description("Action to be taken after processing of the XML file. " +
+      "Possible actions are - " +
+      "1. Delete from the HDFS." +
+      "2. Archived to the target location." +
+      "3. Moved to the target location.")
+    private final String actionAfterProcess;
```

Comment:
I saw a comment on the design doc as well. This does not seem like a configuration that should be related to an XML Reader plugin. It could be a post-action like @anwar6953 suggested on the design doc.